ring/lifecycler.go

```diff
@@ -763,6 +782,14 @@ func (i *Lifecycler) autoJoin(ctx context.Context, targetState InstanceState) er
 			ringDesc = in.(*Desc)
 		}
 
+		conditionalTokenGenerator, ok := i.tokenGenerator.(ConditionalTokenGenerator)
```

Comment:
Some random thoughts:
1. When using memberlist, you don't have the guarantee that all previous instances registration has been propagated yet. Handling this case is complicated and possibly error prone. I'm OK not checking it, but you should add a comment about it, saying that we're aware of it and what we do here is just a best effort.
2. What happens if, for any reason, the previous instances are never registered? This will loop forever. We should probably have a timeout (e.g. if the condition is not honored within 5m, then move on and register anyway).
3. Each time this check fails, we return error to `CAS()` and the KVStore client will retry. However, specific implementations may or may not have a backup (e.g. could retry continuously in a loop, with no throttling) and also specific implementations may have a max number of retries after which the CAS() will just stop and fail. For example, in case of memberlist client, we retry up to 10 times and them we give up. As you can see, this behaviour is incompatible with the new logic you're introducing.

For the reasons above, I believe you should move the logic out of `CAS()`. Before the `CAS()` you could have a loop (with backoff and max retries) to wait until the condition is satified. Then, once the condition is satisfied, we call CAS() as of today.

Reply:
I have added 2 flags: `canJoinEnabled` and `canJoinTimeout` that will be passed to `NewSpreadMinimizingTokenGenerator()` directly from `mimir`.

I have placed the loop checking the previous instances inside `SpreadMinimizingTokenGenerator.CanJoin()` because this way both config flags remain on `SpreadMinimizingTokenGenerator` and other `dskit` sources will not need to know about their existence.